core/src/identity.rs

```diff
@@ -104,6 +104,20 @@ impl Keypair {
         }
     }
 
+    /// Sign a message hash. This function removes the hashing logic from the signing. This
+    /// is identical to sign for Ed25519 and Rsa but requires a 256-bit message (hash) for
+    /// Secp256k1 keys.
+    pub fn sign_raw(&self, msg: &[u8]) -> Result<Vec<u8>, SigningError> {
```

Comment:
I'm not sure I understand this. 

If this was implemented as you suggest, i.e. that `identity::Keypair::sign()` uses `sign_raw()` then the API would not work uniformly for all key types. Because `sign_raw()` (and therefore `identity::Keypair::sign()`) will fail for length's that are not 32 bytes, for `secp256k1`. 

Changing the `identity::Keypair` and high level API to use `sign_raw()` for `secp256k1` keys would be a breaking change for implementations that are expecting a `Sha256` hash in the signature.

The current implementation is uniform across all key types. User's can use `sign()` as they always have and user's who wish to implement their own hash function can use `sign_raw()` (regardless of which keypair you have). It will succeed for all keypairs.

Reply:
To put another way, if you want to sign a 32-byte hash, you can use `sign_raw()` (for all keys). If you have arbitrary data and just want a signature, you can use (`sign()`). 

If you have arbitrary data, you should not use `sign_raw()` (it will fail for `secp256k1` keys, this is why `sign()` still exists). `sign_raw()` may be better named to `sign_hash()`.

Reply:
I had considered `sign_raw(&self, msg: &[u8;32])`, but decided against it due to the extra effort in copying bytes to fixed-size arrays.

Reply:
> I'm not sure I understand this.
> 
> If this was implemented as you suggest, i.e. that `identity::Keypair::sign()` uses `sign_raw()`  [..]

That was not my suggestion. I made that comment in the context of the `identity::secp256k1` module. I'm saying that `sign_raw` and `verify_raw` should not exist in the `identity` module, only in `identity::secp256k1`.

Reply:
Oh yep, understand. 

My specific problem is a protocol which uses a keccak hash for the message. I need to generate a 'Keypair' and sign/verify the message (including those from other implementations). The keypair could be anything provided by libp2p. 

My goal was to have some mechanism that allows me to 'PublicKey:: verify ()' without having to match and perform specific logic if 'secp256k1'. (Specifically, if 'secp256k1' I need to encode and decode the key (into the 'secp256k1::Secp256k1' struct (as the key is private) to perform the fundamental verify without taking the 'Sha256' hash. 

Currently for 'secp256k1' we can only verify messages that have been 'Sha256' hashed. 

Do you have any suggestions for this? 

I'm happy to write the added logic for my specific problem also :)

Reply:
Sorry, but I do not yet understand the problems you're seeing. If you have `identity::secp256k1::PublicKey::verify_raw` and `identity::secp256k1::Keypair::sign_raw`, what stops you from doing what you want to do? If you have a `identity::Keypair` or `identity::PublicKey` you can pattern-match on them and decide to use the `*_raw` variants for secp256k1 keys if you know you have a 32 byte message hash, and delegate to the regular `sign` / `verify` for the other key types, i.e. pretty much what you've done here for `identity::Keypair::sign_raw` and `identity::PublicKey::verify_raw`. I just don't think providing these latter functions as part of `libp2p-core` is a good idea, because of the differing semantics across key types, i.e. for secp256k1 keys it will take the message as is, and for the others it will actually re-hash them. There is simply no equivalent to the secp256k1 `sign_raw` / `verify_raw` for the other key types and hence I think these should not be part of the top-level API.

Reply:
Ah, sorry, I've figured out the confusion. I was under the impression the current `identity::Keypair::sign()` function for `rsa` and `ed25519` didn't perform SHA256 hashes and only `secp256k1` did. I was originally under the impression the `sign()` was not uniform across the types. My bad. 

Definitely agree interface should remain uniform. As you mention, I can solve my problem by matching on `secp256k1` and am happy to do so, however still think it might be useful to have a uniform high level function that removes the fixed in-built hashing. 

I've noticed that `ed25519` allows for this: https://docs.rs/ed25519-dalek/1.0.0-pre.1/ed25519_dalek/struct.Keypair.html (see the `sign_prehashed()` function), however the `rsa` library at first glance requires a digest function as input (perhaps one that leaves the input un-touched). If the `rsa` library accommodated this, what are your thoughts about the rename of `sign_raw()` to `sign_prehashed()` which calls the equivalent functionality on all underlying key types?

Finally, while I'm on the topic, I'm dealing with a spec that uses key-value to store public keys. i.e key: `secp256k1` value: <serialized_public_key>. What are you thoughts about implementing `Into<String> for PublicKey` to generate a string to use as keys (in my particular case). I've currently left it out of `libp2p_core`, curious if you think would be useful for anyone else.

Reply:
> If the rsa library accommodated this, what are your thoughts about the rename of sign_raw() to sign_prehashed() which calls the equivalent functionality on all underlying key types?

The naming sounds good, but even if all key types currently supported by `libp2p-core` support signing and verifying prehashed messages, it would probably still not be a good idea to add such functions to the high-level API (i.e. the top-level `identity` module) for two reasons:

  1. The different signature algorithms require hash functions with different output lengths, which would again make a high-level API a very leaky abstraction.

  2. Any additions to the high-level API are a commitment to functionality that must be supported on all supported key types, including possible future key types. There is a lot of value in keeping the API surface of abstractions like `identity::Keypair` and `identity::PublicKey` small.

There may also be considerations or restrictions regarding which hash algorithms can safely be combined with which signature schemes, I'm not sure.

But adding `sign_prehashed` / `verify_prehashed` to the individual key-type specific submodules for whichever key types support such functionality still sounds good to me.

> Finally, while I'm on the topic, I'm dealing with a spec that uses key-value to store public keys. i.e key: secp256k1 value: <serialized_public_key>. What are you thoughts about implementing Into<String> for PublicKey to generate a string to use as keys (in my particular case). I've currently left it out of libp2p_core, curious if you think would be useful for anyone else.

I think that would be somewhat too ad-hoc. We mainly want to support parsing and serialising keys to and from formats that are widely-used standards or de-facto standards (e.g. as defined in PKCS or the libp2p specs).

Reply:
Yep cool. I agree. Sounds like I'll implement a hash-specific version for my needs. Cheers